add IPFS plugin

davidak opened the pull request:
with Bandwidth and Peers chart

![bildschirmfoto 2016-08-06 um 17 44 59](https://cloud.githubusercontent.com/assets/91113/17457675/0209a386-5bff-11e6-82aa-d4d5b06d23e2.png)

davidak commented:
@paulfantom is this the right way to query multiple URLs?

anything else to improve?

ktsaou commented:
For python guide lines @paulfantom will review.
For sure it needs a conf file with proper defaults, like the rest of the python modules.
Also, if there something to configure on the ipfs side to allow or enable this, please give in the conf file proper instructions.

davidak commented:
- why is the name `ipfs_local` and `postfix: local` different? i think it's cleaner with `:`
- `url` from config don't get passed through to the plugin, it uses default

paulfantom commented on python.d/ipfs.chart.py:
```diff
@@ -0,0 +1,96 @@
+        if len(self.url) == 0:
+            self.baseurl = "http://localhost:5001/"
+        else:
+            self.baseurl = self.url
```
Those lines won't have any meaning since `self.url` will be overridden when `python.d.plugin` executes `check` method (from UrlService).

paulfantom commented:
> why is the name ipfs_local and postfix: local different? i think it's cleaner with :

This is something to be done in `dashboard.html`. If you want I can change it later.

> `url` from config don't get passed through to the plugin, it uses default

That's because you don't use `url`, you use `baseurl`. Configuration is loaded in `check` method, not in constructor (`__init__`).
I think you could add something like this to `Service` class:
```py
def check(self):
    if UrlService.check(self):
        return True
    if len(self.url) == 0:
        self.baseurl = "http://localhost:5001/"
     else:
        self.baseurl = self.url
    test = self._get_data()
    if test is None or len(test) == 0:
        return False
    else:
        return True
```

Some quick explanation, how it works (since documentation is non-existent):
Module needs to have `check`, `create`, and `update` methods, since they are executed by `python.d.plugin`.
Those 3 methods are provided by `SimpleService` class, which also implements many other useful things (like threading or wrappers on chart creation).
Other `*Service` classes are children of `SimpleService` class, and they just provide some helper functions (like HTTP basic auth in `UrlService` or handling socket connections in `SocketService`).

And every `SimpleService` child class needs to have:

1. constructor with:
  * `ParentService.__init__(self, configuration=configuration, name=name)`
  * `self.definitions` - this stores all chart definitions (usually defined in global variable CHARTS)
  * `self.order` - order in which charts are displayed (usually in global variable ORDER)

2. `_get_data` method which returns dictionary where `key` is dimension name and `value` is a number to show on graph.

And if you want to get something from configuration, you need to parse `configuration` variable which stores everything that is defined in yaml file. You can do it either in constructor or in custom `check` method (preferred).
But remember that by default every service parses it's configuration file not in constructor, but in `check` method.

paulfantom commented:
I hope my explanations are good enough.

paulfantom commented on python.d/ipfs.chart.py:
```diff
@@ -0,0 +1,96 @@
+        peers = self._get_peers()
+        bandwidth_in, bandwidth_out = self._get_bandwidth()
+
+        return {'peers': peers, 'in': bandwidth_in, 'out': bandwidth_out}
```
This shouldn't return `None` as dictionary value. If dimension has `None` value it shouldn't be returned.

davidak replied:
should it just `return None` then?

paulfantom replied:
It should check if `peers` and bandwidths are `None`.
If `peers` is `None`, but bandwidths have some data, it should return `{'in': bandwidth_in, 'out': bandwidth_out}`. if `bandwidth_in` is `None` but other have some data it should return `{'peers': peers, 'out': bandwidth_out}`. And so on.
`None` should be returned if there is no data in any dimension.

paulfantom commented:
Also this will be the first service to query multiple URLs.

davidak commented:
Thanks for the explanations!

I added your `check()` code.

```
    if UrlService.check(self):
        return True
```

leads to `ipfs_localhost cannot find check() function.` so i deleted that lines.

i added a `print` to see if i get the correct url from configuration, but it's empty!

```
    def check(self):
        print(self.url)
        if len(self.url) == 0:
```

so that don't work :(

paulfantom commented:
```py
def check(self):
        print(self.url)
        if len(self.url) == 0:
```

with this you overrode gathering data from configuration, so object won't have `self.url`.

davidak commented:
the second makes it clearer where the url comes from, eg. `self.configuration['url']`.

now i can configure multiple checks!

>Also what is returned by ipfs when you request localhost:5001/api/v0/stats ?

```
curl http://localhost:5001/api/v0/stats
{"Message":"This command can't be called directly. Try one of its subcommands.","Code":1}
```

that is no command.

davidak commented:
last thing is the name in the webui menu.

it would be nice to have `postfix: local` only when there are actually multiple instances, else only `postfix` or the instance name other than `local`.

also all names should be displayed as `postfix: local` and not `postfix_local`.

i tried something like this for ipfs in `index.html`, also restartet netdata but nothing changed.

https://github.com/firehol/netdata/commit/be51843cb7749104721f2af1793e34b82e7c6498#diff-d23fe9001e474c96419c8f2519cbef67

@paulfantom can you do that?

paulfantom commented:
To display `ipfs: local` instead of `ipfs_local` on dashboard you need to add two things to index.html:
first is:
```js
case 'redis':
case 'postfix':
case 'redis':
```
and the second:
```js
'redis': {
	title: 'Redis',
	info: undefined
},
'ipfs': {
	title: 'ipfs',
	info: undefined
},
'phpfpm': {
	title: 'PHP-FPM',
	info: undefined,
},
```
And that's all. Remember about refreshing page with Ctrl + F5.

davidak commented:
i have done exactly that, but in `/root/netdata/web/index.html`, not in `/usr/share/netdata/web/index.html`.

so it's ready to merge i thing. :)

@paulfantom thanks again for your help!

paulfantom commented:
Ok, I think it is good.
Merging.
